File: scheduling/scheduling-annotation-processor/src/main/java/io/koraframework/scheduling/annotation/processor/JdkSchedulingGenerator.java
package io.koraframework.scheduling.annotation.processor;

import com.palantir.javapoet.*;
import io.koraframework.annotation.processor.common.*;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.util.Elements;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.Map;
import java.util.Optional;

public class JdkSchedulingGenerator {
    public static ClassName scheduleAtFixedRate = ClassName.get("io.koraframework.scheduling.jdk.annotation", "ScheduleAtFixedRate");
    public static ClassName scheduleOnce = ClassName.get("io.koraframework.scheduling.jdk.annotation", "ScheduleOnce");
    public static ClassName scheduleWithFixedDelay = ClassName.get("io.koraframework.scheduling.jdk.annotation", "ScheduleWithFixedDelay");
    public static ClassName scheduleWithCron = ClassName.get("io.koraframework.scheduling.jdk.annotation", "ScheduleWithCron");

    private static final ClassName fixedDelayJobClassName = ClassName.get("io.koraframework.scheduling.jdk", "FixedDelayJob");
    private static final ClassName fixedRateJobClassName = ClassName.get("io.koraframework.scheduling.jdk", "FixedRateJob");
    private static final ClassName runOnceJobClassName = ClassName.get("io.koraframework.scheduling.jdk", "RunOnceJob");
    private static final ClassName cronJobClassName = ClassName.get("io.koraframework.scheduling.jdk", "CronJob");
    private static final ClassName cronExpressionClassName = ClassName.get("io.koraframework.scheduling.jdk", "CronExpression");
    private static final ClassName schedulingJobConfigClassName = ClassName.get("io.koraframework.scheduling.common", "SchedulingJobConfig");
    private static final ClassName jobTelemetryConfigClassName = ClassName.get("io.koraframework.scheduling.common", "SchedulingJobConfig", "JobTelemetryConfig");
    private static final ClassName schedulingTelemetryFactoryClassName = ClassName.get("io.koraframework.scheduling.common.telemetry", "SchedulingTelemetryFactory");
    private static final ClassName jdkSchedulingExecutor = ClassName.get("io.koraframework.scheduling.jdk", "SchedulingJdkExecutor");
    private final Elements elements;
    private final ProcessingEnvironment processingEnv;

    public JdkSchedulingGenerator(ProcessingEnvironment processingEnv) {
        this.elements = processingEnv.getElementUtils();
        this.processingEnv = processingEnv;
    }

    public void generate(TypeElement type, Element method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var triggerTypeName = ClassName.get((TypeElement) trigger.triggerAnnotation().getAnnotationType().asElement());
        if (triggerTypeName.equals(scheduleAtFixedRate)) {
            this.generateScheduleAtFixedRate(type, method, module, trigger);
            return;
        }
        if (triggerTypeName.equals(scheduleWithFixedDelay)) {
            this.generateScheduleWithFixedDelay(type, method, module, trigger);
            return;
        }
        if (triggerTypeName.equals(scheduleOnce)) {
            this.generateScheduleOnce(type, method, module, trigger);
            return;
        }
        if (triggerTypeName.equals(scheduleWithCron)) {
            this.generateScheduleWithCron(type, method, module, trigger);
            return;
        }
        throw new IllegalStateException("Unknown trigger type: " + trigger.triggerAnnotation().getAnnotationType());
    }

    private void generateScheduleWithCron(TypeElement type, Element method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var configName = AnnotationUtils.<String>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "config");
        var configClassName = NameUtils.generatedType(type, method.getSimpleName() + "_Config");
        var jobMethodName = NameUtils.generatedType(type, method.getSimpleName() + "_Job");
        var cron = AnnotationUtils.<String>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "value");
        var componentMethod = MethodSpec.methodBuilder(jobMethodName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(jdkSchedulingExecutor, "service")
            .addParameter(ParameterizedTypeName.get(CommonClassNames.valueOf, TypeName.get(type.asType())), "object")
            .returns(cronJobClassName)
            .addAnnotation(CommonClassNames.root);

        if (configName == null || configName.isBlank()) {
            if (cron == null || cron.isBlank()) {
                throw new ProcessingErrorException("Either value() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
            }
            componentMethod
                .addStatement("var telemetry = telemetryFactory.get(null, null, $T.class, $S)", type, method.getSimpleName())
                .addStatement("var cron = $T.parse($S)", cronExpressionClassName, cron);
        } else {
            var config = TypeSpec.interfaceBuilder(configClassName)
                .addOriginatingElement(method)
                .addAnnotation(AnnotationUtils.generated(JdkSchedulingGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(schedulingJobConfigClassName)
                .addAnnotation(CommonClassNames.configValueExtractorAnnotation);

            if (cron == null || cron.isBlank()) {
                config.addMethod(MethodSpec.methodBuilder("cron")
                    .returns(String.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                    .build()
                );
            } else {
                config.addMethod(MethodSpec.methodBuilder("cron")
                    .returns(String.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .addStatement("return $S", cron)
                    .build()
                );
            }

            module.addMethod(cronConfigComponent(packageName, configClassName, configName, cron));
            CommonUtils.safeWriteTo(this.processingEnv, JavaFile.builder(packageName, config.build()).build());

            componentMethod.addParameter(ClassName.get(packageName, configClassName), "config");
            componentMethod.addStatement("var telemetry = telemetryFactory.get($S, config.telemetry(), $T.class, $S)", configName, type, method.getSimpleName());
            componentMethod.addStatement("var cron = $T.parse(config.cron())", cronExpressionClassName);
        }

        componentMethod.addStatement("return new $T(telemetry, service, () -> object.get().$N(), cron)", cronJobClassName, method.getSimpleName());
        module.addMethod(componentMethod.build());
    }

    private void generateScheduleOnce(TypeElement type, Element method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var configName = AnnotationUtils.<String>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "config");
        var configClassName = NameUtils.generatedType(type, method.getSimpleName() + "_Config");
        var jobMethodName = NameUtils.generatedType(type, method.getSimpleName() + "_Job");
        var delay = AnnotationUtils.<Long>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "delay");
        var unit = AnnotationUtils.<VariableElement>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "unit");
        var componentMethod = MethodSpec.methodBuilder(jobMethodName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(jdkSchedulingExecutor, "service")
            .addParameter(ParameterizedTypeName.get(CommonClassNames.valueOf, TypeName.get(type.asType())), "object")
            .returns(runOnceJobClassName)
            .addAnnotation(CommonClassNames.root);

        if (configName.isEmpty()) {
            if (delay == null || delay == 0) {
                throw new ProcessingErrorException("Either delay() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
            }
            componentMethod
                .addStatement("var telemetry = telemetryFactory.get($S, null, $T.class, $S)", type.getQualifiedName() + "#" + method.getSimpleName(), type, method.getSimpleName())
                .addStatement("var delay = $T.of($L, $T.$L)", Duration.class, delay, ChronoUnit.class, unit);
        } else {
            var config = TypeSpec.interfaceBuilder(configClassName)
                .addOriginatingElement(method)
                .addAnnotation(AnnotationUtils.generated(JdkSchedulingGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(schedulingJobConfigClassName)
                .addAnnotation(CommonClassNames.configValueExtractorAnnotation);

            if (delay == null || delay == 0) {
                config.addMethod(MethodSpec.methodBuilder("delay")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                    .build()
                );
            } else {
                config.addMethod(MethodSpec.methodBuilder("delay")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .addStatement("return $T.of($L, $T.$L)", Duration.class, delay, ChronoUnit.class, unit)
                    .build()
                );
            }
            module.addMethod(configComponent(packageName, configClassName, configName));
            CommonUtils.safeWriteTo(this.processingEnv, JavaFile.builder(packageName, config.build()).build());

            componentMethod.addParameter(ClassName.get(packageName, configClassName), "config");
            componentMethod.addStatement("var telemetry = telemetryFactory.get($S, config.telemetry(), $T.class, $S)", configName, type, method.getSimpleName());
            componentMethod.addStatement("var delay = config.delay()");
        }

        componentMethod.addStatement("return new $T(telemetry, service, () -> object.get().$N(), delay)", runOnceJobClassName, method.getSimpleName());
        module.addMethod(componentMethod.build());
    }

    private void generateScheduleWithFixedDelay(TypeElement type, Element method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var configName = AnnotationUtils.<String>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "config");
        var configClassName = NameUtils.generatedType(type, method.getSimpleName() + "_Config");
        var jobMethodName = NameUtils.generatedType(type, method.getSimpleName() + "_Job");
        var initialDelay = AnnotationUtils.<Long>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "initialDelay");
        var delay = AnnotationUtils.<Long>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "delay");
        var unit = AnnotationUtils.<VariableElement>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "unit");
        var componentMethod = MethodSpec.methodBuilder(jobMethodName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(jdkSchedulingExecutor, "service")
            .addParameter(ParameterizedTypeName.get(CommonClassNames.valueOf, TypeName.get(type.asType())), "object")
            .returns(fixedDelayJobClassName)
            .addAnnotation(CommonClassNames.root);

        if (configName.isEmpty()) {
            if (delay == null || delay == 0) {
                throw new ProcessingErrorException("Either delay() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
            }
            componentMethod
                .addStatement("var telemetry = telemetryFactory.get($S, null, $T.class, $S)", type.getQualifiedName() + "#" + method.getSimpleName(), type, method.getSimpleName())
                .addStatement("var initialDelay = $T.of($L, $T.$L)", Duration.class, initialDelay, ChronoUnit.class, unit)
                .addStatement("var delay = $T.of($L, $T.$L)", Duration.class, delay, ChronoUnit.class, unit);
        } else {
            var config = TypeSpec.interfaceBuilder(configClassName)
                .addOriginatingElement(method)
                .addAnnotation(AnnotationUtils.generated(JdkSchedulingGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(schedulingJobConfigClassName)
                .addAnnotation(CommonClassNames.configValueExtractorAnnotation);

            componentMethod.addParameter(ClassName.get(packageName, configClassName), "config");
            if (delay == null || delay == 0) {
                config.addMethod(MethodSpec.methodBuilder("delay")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                    .build()
                );
            } else {
                config.addMethod(MethodSpec.methodBuilder("delay")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .addStatement("return $T.of($L, $T.$L)", Duration.class, delay, ChronoUnit.class, unit)
                    .build()
                );
            }
            config.addMethod(MethodSpec.methodBuilder("initialDelay")
                .returns(Duration.class)
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .addStatement("return $T.of($L, $T.$L)", Duration.class, initialDelay, ChronoUnit.class, unit)
                .build()
            );
            module.addMethod(configComponent(packageName, configClassName, configName));
            CommonUtils.safeWriteTo(this.processingEnv, JavaFile.builder(packageName, config.build()).build());

            componentMethod
                .addStatement("var telemetry = telemetryFactory.get($S, config.telemetry(), $T.class, $S)", configName, type, method.getSimpleName())
                .addStatement("var initialDelay = config.initialDelay()")
                .addStatement("var delay = config.delay()");
        }
        componentMethod
            .addStatement("return new $T(telemetry, service, () -> object.get().$N(), initialDelay, delay)", fixedDelayJobClassName, method.getSimpleName());
        module.addMethod(componentMethod.build());
    }

    private void generateScheduleAtFixedRate(TypeElement type, Element method, TypeSpec.Builder module, SchedulingTrigger trigger) {
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var configName = AnnotationUtils.<String>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "config");
        var configClassName = NameUtils.generatedType(type, method.getSimpleName() + "_Config");
        var jobMethodName = NameUtils.generatedType(type, method.getSimpleName() + "_Job");
        var initialDelay = AnnotationUtils.<Long>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "initialDelay");
        var period = AnnotationUtils.<Long>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "period");
        var unit = AnnotationUtils.<VariableElement>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "unit");
        var componentMethod = MethodSpec.methodBuilder(jobMethodName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(jdkSchedulingExecutor, "service")
            .addParameter(ParameterizedTypeName.get(CommonClassNames.valueOf, TypeName.get(type.asType())), "object")
            .returns(fixedRateJobClassName)
            .addAnnotation(CommonClassNames.root);

        if (configName.isEmpty()) {
            if (period == null || period == 0) {
                throw new ProcessingErrorException("Either period() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
            }
            componentMethod
                .addStatement("var telemetry = telemetryFactory.get($S, null, $T.class, $S)", type.getQualifiedName() + "#" + method.getSimpleName(), type, method.getSimpleName())
                .addStatement("var initialDelay = $T.of($L, $T.$L)", Duration.class, initialDelay, ChronoUnit.class, unit)
                .addStatement("var period = $T.of($L, $T.$L)", Duration.class, period, ChronoUnit.class, unit);
        } else {
            var config = TypeSpec.interfaceBuilder(configClassName)
                .addOriginatingElement(method)
                .addAnnotation(AnnotationUtils.generated(JdkSchedulingGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(schedulingJobConfigClassName)
                .addAnnotation(CommonClassNames.configValueExtractorAnnotation);
            if (period == null || period == 0) {
                config.addMethod(MethodSpec.methodBuilder("period")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
                    .build()
                );
            } else {
                config.addMethod(MethodSpec.methodBuilder("period")
                    .returns(Duration.class)
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .addStatement("return $T.of($L, $T.$L)", Duration.class, period, ChronoUnit.class, unit)
                    .build()
                );
            }
            config.addMethod(MethodSpec.methodBuilder("initialDelay")
                .returns(Duration.class)
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .addStatement("return $T.of($L, $T.$L)", Duration.class, initialDelay, ChronoUnit.class, unit)
                .build()
            );
            module.addMethod(configComponent(packageName, configClassName, configName));
            CommonUtils.safeWriteTo(this.processingEnv, JavaFile.builder(packageName, config.build()).build());

            componentMethod.addParameter(ClassName.get(packageName, configClassName), "config");
            componentMethod
                .addStatement("var telemetry = telemetryFactory.get($S, config.telemetry(), $T.class, $S)", configName, type, method.getSimpleName())
                .addStatement("var initialDelay = config.initialDelay()")
                .addStatement("var period = config.period()");
        }
        componentMethod
            .addStatement("return new $T(telemetry, service, () -> object.get().$N(), initialDelay, period)", fixedRateJobClassName, method.getSimpleName());
        module.addMethod(componentMethod.build());
    }

    private static MethodSpec configComponent(String packageName, String configClassName, String configPath) {
        return MethodSpec.methodBuilder(configClassName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(CommonClassNames.config, "config")
            .addParameter(
                ParameterizedTypeName.get(
                    ClassName.get("io.koraframework.config.common.extractor", "ConfigValueExtractor"),
                    ClassName.get(packageName, configClassName)
                ),
                "extractor"
            )
            .addStatement("return extractor.extractOrThrow(config.get($S))", configPath)
            .returns(ClassName.get(packageName, configClassName))
            .build();

    }

    private static MethodSpec cronConfigComponent(String packageName, String configClassName, String configPath, String defaultCron) {
        var configType = ClassName.get(packageName, configClassName);
        var method = MethodSpec.methodBuilder(configClassName)
            .addModifiers(Modifier.DEFAULT, Modifier.PUBLIC)
            .addParameter(CommonClassNames.config, "config")
            .addParameter(ParameterizedTypeName.get(CommonClassNames.configValueExtractor, configType), "extractor")
            .returns(configType)
            .addStatement("var value = config.get($S)", configPath);

        if (defaultCron != null && !defaultCron.isBlank()) {
            method.beginControlFlow("if (value instanceof $T.NullValue)", CommonClassNames.configValue)
                .addCode("return extractor.extractOrThrow($>\n")
                .addCode("new $T.ObjectValue(value.origin(), $T.of($S, new $T.StringValue(value.origin(), $S)))", CommonClassNames.configValue, Map.class, "cron", CommonClassNames.configValue, defaultCron)
                .addCode("$<\n);\n")
                .endControlFlow();
        }
        method.beginControlFlow("if (value instanceof $T.StringValue str)", CommonClassNames.configValue)
            .addStatement("var cron = str.value()")
            .addCode("return extractor.extractOrThrow($>\n")
            .addCode("new $T.ObjectValue(value.origin(), $T.of($S, new $T.StringValue(value.origin(), cron)))", CommonClassNames.configValue, Map.class, "cron", CommonClassNames.configValue)
            .addCode("$<\n);\n")
            .nextControlFlow("else if (value instanceof $T.ObjectValue obj)", CommonClassNames.configValue)
            .addStatement("return extractor.extractOrThrow(obj)")
            .nextControlFlow("else")
            .addStatement("throw $T.unexpectedValueType(value, $T.StringValue.class)", CommonClassNames.configValueExtractionException, CommonClassNames.configValue)
            .endControlFlow();
        return method.build();
    }
}

File: scheduling/scheduling-symbol-processor/src/main/kotlin/io/koraframework/scheduling/symbol/processor/JdkSchedulingGenerator.kt
package io.koraframework.scheduling.symbol.processor

import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.writeTo
import io.koraframework.ksp.common.AnnotationUtils.findEnumValue
import io.koraframework.ksp.common.AnnotationUtils.findValue
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.KotlinPoetUtils.controlFlow
import io.koraframework.ksp.common.KspCommonUtils.generated
import io.koraframework.ksp.common.exception.ProcessingErrorException
import io.koraframework.ksp.common.getOuterClassesAsPrefix
import java.time.Duration

class JdkSchedulingGenerator(val environment: SymbolProcessorEnvironment) {
    private val fixedDelayJobClassName = ClassName("io.koraframework.scheduling.jdk", "FixedDelayJob")
    private val fixedRateJobClassName = ClassName("io.koraframework.scheduling.jdk", "FixedRateJob")
    private val runOnceJobClassName = ClassName("io.koraframework.scheduling.jdk", "RunOnceJob")
    private val cronJobClassName = ClassName("io.koraframework.scheduling.jdk", "CronJob")
    private val cronExpressionClassName = ClassName("io.koraframework.scheduling.jdk", "CronExpression")
    private val jdkSchedulingExecutor = ClassName("io.koraframework.scheduling.jdk", "SchedulingJdkExecutor")
    private val schedulingTelemetryFactoryClassName = ClassName("io.koraframework.scheduling.common.telemetry", "SchedulingTelemetryFactory")
    private val schedulingJobConfigClassName = ClassName("io.koraframework.scheduling.common", "SchedulingJobConfig")
    private val jobTelemetryConfigClassName = ClassName("io.koraframework.scheduling.common", "SchedulingJobConfig", "JobTelemetryConfig")

    fun generate(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        when (trigger.annotation.shortName.asString()) {
            "ScheduleAtFixedRate" -> this.generateScheduleAtFixedRate(type, function, builder, trigger)
            "ScheduleWithFixedDelay" -> this.generateScheduleWithFixedDelay(type, function, builder, trigger)
            "ScheduleOnce" -> this.generateScheduleOnce(type, function, builder, trigger)
            "ScheduleWithCron" -> this.generateScheduleWithCron(type, function, builder, trigger)
        }
    }

    private fun generateScheduleWithCron(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val packageName = type.packageName.asString()
        val configName = trigger.annotation.findValue<String>("config")
        val typeClassName = type.toClassName()
        val jobFunName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job"
        val cron = trigger.annotation.findValue<String>("value")
        val componentFunction = FunSpec.builder(jobFunName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("service", jdkSchedulingExecutor)
            .addParameter("target", CommonClassNames.valueOf.parameterizedBy(typeClassName))
            .returns(cronJobClassName)
            .addAnnotation(CommonClassNames.root)

        if (configName.isNullOrBlank()) {
            if (cron.isNullOrBlank()) {
                throw ProcessingErrorException("Either value() or config() annotation parameter must be provided", function)
            }
            componentFunction
                .addStatement("val telemetry = telemetryFactory.get(null, null, %T::class.java, %S)", typeClassName, function.simpleName.getShortName())
                .addStatement("val cron = %T.parse(%S)", cronExpressionClassName, cron)
        } else {
            val configType = cronConfigType(type, function, cron ?: "")
            FileSpec.get(packageName, configType).writeTo(environment.codeGenerator, false, listOf(type.containingFile!!))

            componentFunction
                .addParameter("config", ClassName(packageName, configType.name!!))
                .addStatement("val telemetry = telemetryFactory.get(%S, config.telemetry(), %T::class.java, %S)", configName, typeClassName, function.simpleName.getShortName())
                .addStatement("val cron = %T.parse(config.cron())", cronExpressionClassName)
            builder.addFunction(cronConfigComponent(packageName, configType.name!!, configName, cron ?: ""))
        }
        componentFunction
            .addStatement("return %T(telemetry, service, { target.get().%N() }, cron)", cronJobClassName, function.simpleName.getShortName())
        builder.addFunction(componentFunction.build())
    }


    private fun generateScheduleAtFixedRate(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val packageName = type.packageName.asString()
        val configName = trigger.annotation.findValue<String>("config")
        val typeClassName = type.toClassName()
        val jobFunName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job"
        val initialDelay = trigger.annotation.findValue<Long>("initialDelay") ?: 0
        val period = trigger.annotation.findValue<Long>("period")
        val unit = trigger.annotation.findEnumValue("unit")!!
        val componentFunction = FunSpec.builder(jobFunName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("service", jdkSchedulingExecutor)
            .addParameter("target", CommonClassNames.valueOf.parameterizedBy(typeClassName))
            .returns(fixedRateJobClassName)
            .addAnnotation(CommonClassNames.root)

        if (configName.isNullOrBlank()) {
            if (period == null || period == 0L) {
                throw ProcessingErrorException("Either period() or config() annotation parameter must be provided", function)
            }
            componentFunction
                .addStatement("val initialDelay = %T.of(%L, %L)", Duration::class, initialDelay, unit)
                .addStatement("val period = %T.of(%L, %L)", Duration::class, period, unit)
                .addStatement("val telemetry = telemetryFactory.get(null, null, %T::class.java, %S)", typeClassName, function.simpleName.getShortName())
        } else {
            val configType = configType(
                type, function,
                ConfigParameter("period", Duration::class.asClassName(), period?.let { CodeBlock.of("%T.of(%L, %L)", Duration::class, it, unit) }),
                ConfigParameter("initialDelay", Duration::class.asClassName(), CodeBlock.of("%T.of(%L, %L)", Duration::class, initialDelay, unit)),
            )
            FileSpec.get(packageName, configType).writeTo(environment.codeGenerator, false, listOf(type.containingFile!!))

            componentFunction
                .addParameter("config", ClassName(packageName, configType.name!!))
                .addStatement("val telemetry = telemetryFactory.get(%S, config.telemetry(), %T::class.java, %S)", configName, typeClassName, function.simpleName.getShortName())
                .addStatement("val period = config.period()")
                .addStatement("val initialDelay = config.initialDelay()")
            builder.addFunction(configComponent(packageName, configType.name!!, configName))
        }
        componentFunction
            .addStatement("return %T(telemetry, service, { target.get().%N() }, initialDelay, period)", fixedRateJobClassName, function.simpleName.getShortName())
        builder.addFunction(componentFunction.build())
    }

    private fun generateScheduleWithFixedDelay(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val packageName = type.packageName.asString()
        val configName = trigger.annotation.findValue<String>("config")
        val typeClassName = type.toClassName()
        val jobFunName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job"
        val initialDelay = trigger.annotation.findValue<Long>("initialDelay") ?: 0
        val delay = trigger.annotation.findValue<Long>("delay")
        val unit = trigger.annotation.findEnumValue("unit")!!
        val componentFunction = FunSpec.builder(jobFunName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("service", jdkSchedulingExecutor)
            .addParameter("target", CommonClassNames.valueOf.parameterizedBy(typeClassName))
            .returns(fixedDelayJobClassName)
            .addAnnotation(CommonClassNames.root)

        if (configName.isNullOrBlank()) {
            if (delay == null || delay == 0L) {
                throw ProcessingErrorException("Either delay() or config() annotation parameter must be provided", function)
            }
            componentFunction
                .addStatement("val telemetry = telemetryFactory.get(null, null, %T::class.java, %S)", typeClassName, function.simpleName.getShortName())
                .addStatement("val initialDelay = %T.of(%L, %L)", Duration::class, initialDelay, unit)
                .addStatement("val delay = %T.of(%L, %L)", Duration::class, delay, unit)
        } else {
            val configType = configType(
                type, function,
                ConfigParameter("delay", Duration::class.asClassName(), delay?.let { CodeBlock.of("%T.of(%L, %L)", Duration::class, it, unit) }),
                ConfigParameter("initialDelay", Duration::class.asClassName(), CodeBlock.of("%T.of(%L, %L)", Duration::class, initialDelay, unit)),
            )
            FileSpec.get(packageName, configType).writeTo(environment.codeGenerator, false, listOf(type.containingFile!!))

            componentFunction
                .addParameter("config", ClassName(packageName, configType.name!!))
                .addStatement("val telemetry = telemetryFactory.get(%S, config.telemetry(), %T::class.java, %S)", configName, typeClassName, function.simpleName.getShortName())
                .addStatement("val delay = config.delay()")
                .addStatement("val initialDelay = config.initialDelay()")
            builder.addFunction(configComponent(packageName, configType.name!!, configName))
        }
        componentFunction
            .addStatement("return %T(telemetry, service, { target.get().%N() }, initialDelay, delay)", fixedDelayJobClassName, function.simpleName.getShortName())
        builder.addFunction(componentFunction.build())
    }

    private fun generateScheduleOnce(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val packageName = type.packageName.asString()
        val configName = trigger.annotation.findValue<String>("config")
        val typeClassName = type.toClassName()
        val jobFunName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job"
        val delay = trigger.annotation.findValue<Long>("delay")
        val unit = trigger.annotation.findEnumValue("unit")!!
        val componentFunction = FunSpec.builder(jobFunName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("service", jdkSchedulingExecutor)
            .addParameter("target", CommonClassNames.valueOf.parameterizedBy(typeClassName))
            .returns(runOnceJobClassName)
            .addAnnotation(CommonClassNames.root)

        if (configName.isNullOrBlank()) {
            if (delay == null || delay == 0L) {
                throw ProcessingErrorException("Either delay() or config() annotation parameter must be provided", function)
            }
            componentFunction
                .addStatement("val telemetry = telemetryFactory.get(null, null, %T::class.java, %S)", typeClassName, function.simpleName.getShortName())
                .addStatement("val delay = %T.of(%L, %L)", Duration::class, delay, unit)
        } else {
            val configType = configType(
                type, function,
                ConfigParameter("delay", Duration::class.asClassName(), delay?.let { CodeBlock.of("%T.of(%L, %L)", Duration::class, it, unit) })
            )
            FileSpec.get(packageName, configType).writeTo(environment.codeGenerator, false, listOf(type.containingFile!!))

            componentFunction
                .addParameter("config", ClassName(packageName, configType.name!!))
                .addStatement("val telemetry = telemetryFactory.get(%S, config.telemetry(), %T::class.java, %S)", configName, typeClassName, function.simpleName.getShortName())
                .addStatement("val delay = config.delay()")
            builder.addFunction(configComponent(packageName, configType.name!!, configName))
        }
        componentFunction
            .addStatement("return %T(telemetry, service, { target.get().%N() }, delay)", runOnceJobClassName, function.simpleName.getShortName())
        builder.addFunction(componentFunction.build())
    }

    private fun configComponent(packageName: String, configClassName: String, configPath: String) = FunSpec.builder(configClassName)
        .addParameter("config", CommonClassNames.config)
        .addParameter(
            "extractor", CommonClassNames.configValueExtractor
                .parameterizedBy(ClassName(packageName, configClassName))
        )
        .addStatement("return extractor.extract(config.get(%S))", configPath)
        .returns(ClassName(packageName, configClassName))
        .build()


    private data class ConfigParameter(val name: String, val type: ClassName, val defaultValue: CodeBlock?)

    private fun configType(type: KSClassDeclaration, function: KSFunctionDeclaration, vararg params: ConfigParameter): TypeSpec {
        val configClassName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Config"
        val configType = TypeSpec.interfaceBuilder(configClassName)
            .addAnnotation(CommonClassNames.configValueExtractorAnnotation)
            .generated(JdkSchedulingGenerator::class)
            .addSuperinterface(schedulingJobConfigClassName)
        for (param in params) {
            configType.addFunction(
                FunSpec.builder(param.name)
                    .returns(param.type)
                    .apply {
                        param.defaultValue?.let {
                            addStatement("return %L", it)
                        }
                    }
                    .build()
            )
        }
        return configType.build()
    }

    private fun cronConfigType(type: KSClassDeclaration, function: KSFunctionDeclaration, defaultCron: String): TypeSpec {
        val configClassName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Config"
        val configType = TypeSpec.interfaceBuilder(configClassName)
            .addAnnotation(CommonClassNames.configValueExtractorAnnotation)
            .generated(JdkSchedulingGenerator::class)
            .addSuperinterface(schedulingJobConfigClassName)
        if (defaultCron.isBlank()) {
            configType.addFunction(FunSpec.builder("cron").addModifiers(KModifier.ABSTRACT).returns(STRING).build())
        } else {
            configType.addFunction(FunSpec.builder("cron").returns(STRING).addStatement("return %S", defaultCron).build())
        }
        return configType.build()
    }

    private fun cronConfigComponent(packageName: String, configClassName: String, configPath: String, defaultCron: String) = FunSpec.builder(configClassName)
        .addParameter("config", CommonClassNames.config)
        .addParameter("extractor", CommonClassNames.configValueExtractor.parameterizedBy(ClassName(packageName, configClassName)))
        .addStatement("val value = config.get(%S)", configPath)
        .apply {
            if (defaultCron.isNotBlank()) {
                controlFlow("if (value is %T.NullValue)", CommonClassNames.configValue) {
                    addCode("return extractor.extract(\n")
                    addCode("  %T.ObjectValue(value.origin(), mapOf(%S to %T.StringValue(value.origin(), %S)))\n", CommonClassNames.configValue, "cron", CommonClassNames.configValue, defaultCron)
                    addCode(")!!\n")
                }
            }
        }
        .controlFlow("if (value is %T.ObjectValue)", CommonClassNames.configValue) {
            addStatement("return extractor.extract(value)!!")
        }
        .controlFlow("if (value is %T.StringValue)", CommonClassNames.configValue) {
            addCode("return extractor.extract(\n")
            addCode("  %T.ObjectValue(value.origin(), mapOf(%S to %T.StringValue(value.origin(), value.value())))\n", CommonClassNames.configValue, "cron", CommonClassNames.configValue)
            addCode(")!!\n")
            nextControlFlow("else")
            addStatement(
                "throw %T.unexpectedValueType(value, %T.StringValue::class.java)",
                ClassName("io.koraframework.config.common.extractor", "ConfigValueExtractionException"),
                CommonClassNames.configValue
            )
        }
        .returns(ClassName(packageName, configClassName))
        .build()
}
